baseplate/context/cassandra.py

```diff
@@ -90,15 +97,58 @@ def make_object_for_context(self, name, server_span):
         return cqlmapper.connection.Connection(session_adapter)
 
 
-def _on_execute_complete(_, span):
+class WaitForCallbackResponseFuture(object):
+    """Wrap the ResponseFuture to ensure callbacks have completed.
+
+    This fixes a race condition where the server span can complete before
+    the callback has closed out the child span.
+
+    """
+
+    def __init__(self, span, future, callback_fn, callback_args, errback_fn, errback_args):
+        self.callback_event = Event()
+
+        future.add_callback(callback_fn, callback_args, self.callback_event)
+        future.add_errback(errback_fn, errback_args, self.callback_event)
+
+        self.future = future
+
+    def result(self):
+        exc = None
+        exc_info = None
+
+        try:
+            result = self.future.result()
+        except Exception as e:
+            exc = e
+            exc_info = sys.exc_info()
+
+        # wait for either _on_execute_complete or _on_execute_failed to run
+        wait_result = self.callback_event.wait(timeout=0.01)
+        if not wait_result:
+            logger.warning("callback took too long")
```

Comment:
I think this might be a big vague/confusing of a message to see in logs. If we want to keep it at warning (which sounds reasonable if its presence means we'd be losing metric data) then maybe:

> Cassandra metrics callback took too long. Some metrics may be lost.

? not sure there's much remediation that can be done in this case.